File: Books/A_Comprehensive_Approach_to_ML_and_AI/matlab/eda_tutorial.m
close all; clear all; clc;

% Generate a toy dataset: 200 samples, 5 features
X = randn(200, 5);
k = 2;  % Reduce to 2 dimensions
[Z, W] = simplePCA(X, k);

% Plot the projected data
figure;
scatter(Z(:,1), Z(:,2), 50, 'filled');
title('PCA: First Two Principal Components');
xlabel('Principal Component 1');
ylabel('Principal Component 2');
grid on;

% save_all_figs_OPTION('results/eda','png',1)

function [Z, W] = simplePCA(X, k)
    % simplePCA: Perform PCA on data matrix X to reduce its dimensionality to k.
    % Input:  X (n-by-p data matrix), k (number of principal components)
    % Output: Z (projected data, n-by-k), W (projection matrix, p-by-k)
    
    % Step 1: Center the data
    X_mean = mean(X, 1);
    X_centered = X - X_mean;
    
    % Step 2: Compute the covariance matrix
    S = (1/size(X,1)) * (X_centered' * X_centered);
    
    % Step 3: Eigenvalue decomposition
    [V, D] = eig(S);
    eigenvalues = diag(D);
    
    % Step 4: Sort eigenvectors by decreasing eigenvalues
    [~, idx] = sort(eigenvalues, 'descend');
    V = V(:, idx);
    
    % Step 5: Select the top k eigenvectors
    W = V(:, 1:k);
    
    % Step 6: Project the data
    Z = X_centered * W;
end

File: Books/A_Comprehensive_Approach_to_ML_and_AI/python/eda_tutorial.py
#!/usr/bin/env python3
"""
edaTutorial.py
--------------
This tutorial demonstrates an exploratory data analysis (EDA) pipeline.
A synthetic dataset with 5 features is generated, and PCA is applied to reduce the dimensionality to 2.
The projected data are then visualized.
"""

import numpy as np
import matplotlib.pyplot as plt

np.random.seed(1)
X = np.random.randn(200, 5)  # 200 samples, 5 features

def simple_pca(X, k):
    # Center data
    X_centered = X - np.mean(X, axis=0)
    # Covariance matrix
    S = np.cov(X_centered, rowvar=False)
    # Eigen-decomposition
    eigvals, eigvecs = np.linalg.eigh(S)
    # Sort in descending order
    idx = np.argsort(eigvals)[::-1]
    eigvecs = eigvecs[:, idx]
    # Select top k eigenvectors
    W = eigvecs[:, :k]
    Z = X_centered @ W
    return Z, W

Z, W = simple_pca(X, 2)

plt.figure()
plt.scatter(Z[:,0], Z[:,1], c='b', edgecolors='k')
plt.title("PCA: First Two Principal Components")
plt.xlabel("PC1")
plt.ylabel("PC2")
plt.grid(True)
plt.show()
